File: Programmers/Level 2/Level2_요격시스템.py
targets = [[4,8],[4,5],[10,14],[11,13],[5,12],[3,7],[1,4]]

def solution(targets):
    answer = 0    
    targets.sort()
    print(targets)
    return answer

# 탐욕법을 생각했으나 기준을 잡기가 모호했음.
# 도저히 방법이 감이 잡히지 않아 결국 다른 사람의 풀이 확인
# 그리드 방식을 사용하여 문제 풀이를 진행해야 했었음

missile_Count = 0
s, e = 0, 0

targets.sort(key=lambda x: [x[1], x[0]])
print(targets)

for target in targets:
    # 겹치는 부분을 벗어난 경우 : 겹치지 않기 때문에 무조건 새로운 미사일을 발사해야함.
    if target[0] >= e:
        missile_Count += 1
        e = target[1]
print(missile_Count)

# ** 미사일의 x좌표 범위를 판단할때 시작 지점을 기준으로 생각을 했는데 아니였음.

File: Programmers/Level 2/Level2_요격시스템.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

bool cmp(vector<int>& a, vector<int>& b)
{
    if (a[1] == b[1])
        return a[0] < b[0];
    else
        return a[1] < b[1];
}

int solution(vector<vector<int>> targets)
{
    int answer = 0;
    int endRange = 0;

    sort(targets.begin(), targets.end(), cmp);
    for (int i = 0; i < targets.size(); i++)
    {
        if (targets[i][0] >= endRange)
        {
            answer += 1;
            endRange = targets[i][1];
        }
    }

    return answer;
}

// 1차 시도 : 성공, 끝나는 범위를 기준으로 발사하면 최소 미사일 수가 나오는 간단한 문제